### DISCUSSION TITLE
latency comparison between labwc, Wayfire and X11

### DISCUSSION THREAD

**Discussion by ghollingworth** [labwc/labwc#1810] [General]
One of my team has been building a simple device to measure the latency between a mouse movement and the cursor moving on screen.  Basically he's got a Raspberry Pi Pico which sends a mouse move to the Raspberry Pi and then uses a photodiode to register the on screen movement.

I asked him to compare the performance of X11, Wayfire and labwc with either the GPU renderer or the pixman backend.

X11 - 1 frame (Note there is no compositor running)
Wayfire - 2 frames
labwc - 3 frames

Since both Wayfire and labwc are using the wlroots library, I had assumed they would be the same.  Is there any reason why labwc is taking an extra frame to get a change to the screen?

Gordon

**Comment by Consolatis** [comment]
> with either the GPU renderer or the pixman backend.

Were there differences between the 2?

The exact timing when the mouse movement starts in relation to the last committed frame may be relevant.
Hardware cursors (via DRM cursor plane) or software cursors (composited on top of the already composited frame) may also play a role here.

**Comment by ghollingworth** [comment]
No both GPU and pixman gave the same 3 frame delay, you can see the graph of measurements made (note "labwc" means labwc using the GPU renderer)

I don't think this is related to the hardware cursor because we're using the atomic DRM commits which means the cursor will always appear with the correct frame (annoyingly!)

![image](https://github.com/labwc/labwc/assets/2048540/2d49f96d-3290-485d-a3cb-331435242a21)

**Comment by ghollingworth** [comment]
I would ask David, but he's on holiday this week...  and I was interested...

**Comment by ghollingworth** [comment]
Went back to the person doing it and because of the shape of the results I didn't believe he'd quite got it right.  He had another look and worked out that indeed there was a bug in his latency measurement system.

Now the results give one frame less, so:

X11: averages 14ms, less than one frame (the mouse is using a hardware cursor and is being updated immediately)
Wayfire: averages 25ms, between one and two frames
LabWC: averages 43ms, between two and three frames

So not as bad, but still there's a difference between Wayfire and LabWC.  Do you think the extra frame is caused by the scene graph code (I believe Wayfire doesn't use that but uses it's own implementation)?

**Reply by Consolatis** [reply]
> Do you think the extra frame is caused by the scene graph code (I believe Wayfire doesn't use that but uses it's own implementation)?

AFAIR the cursor is not handled by wlr_scene. However the output commits somewhat are and based of my memory (e.g. may be wrong) the commit for the cursor plane for hardware cursors is linked to the output commit in general. Is there a difference if you use `WLR_NO_HARDWARE_CURSORS=1` to force software cursors? Another option to eliminate the whole cursor topic as potential cause is to have a wayland surface that disables the cursor but instead renders something on cursor movement within its own surface which could then be detected by the photo diode.

You could use the latest sway for comparison as well, that one moved to scene graph.

Overall, although labwc tries to provide decent latency / performance it is not the main goal of labwc. The main goal is more about stability and acceptable resource usage instead. Though that doesn't mean that we wouldn't be interested in fixing whatever causes that additional ~ 1 frame latency but if that would mean unreasonably complicating the codebase and thus making labwc harder to maintain in the future we'd rather live with the additional latency.

**Comment by johanmalm** [comment]
This is very interesting.

I've never spent much time on the rendering chain - have just taken what `wlroots` provides - so can't speak with much intuition on this. 

I'm thinking that unless Wayfire does something fundamentally different (e.g. with respects to buffering or page-flips), then maybe it simply relates to when we render relative to the `present`. For example if we commit a frame very early in the 17 ms period, we might just miss the cursor move in that frame, and if Wayfire is more sophisticated with the frame-scheduling they may just gain 1 frame. (Just a theory though). `sway` has a `max_render_time` config option which can enable very late scheduling. Also, there was some work done by rose in the summer to make this more sophisticated. Not sure how much of it got committed, but might be useful for exploring this further. https://blog.krx.sh/

We do support tearing (for gamers mostly), so that could be useful for gathering data.
https://labwc.github.io/labwc-config.5.html#entry_core_allowtearing

Suggested next steps:
1. Try with `sway` master. (as @Consolatis)  suggested.
2. Use `max_render_time` to schedule late and see if we can save a frame that way.
3. Reach out to Wayfire team to see if they can think of any differences wrt double-buffering and frame-scheduling.

**Comment by droc12345** [comment]
all they all using the same renderer? ie all gles2, pixman, etc

**Comment by RicArch97** [comment]
Would be nice to see how Sway performs compared to the others indeed. Even without async pageflips, Sway always felt really snappy and quite close to X11 in terms of latency.

> Overall, although labwc tries to provide decent latency / performance it is not the main goal of labwc.

I've been using Sway for over 4 years, but i am quite interested in a stacking compositor as well. LabWC seems to tick a lot of boxes, though i wouldn't feel comfortable switching over knowing that I'd have to deal with increased latency in games. Since gaming is actually viable on Linux these days, performance may now be way more important to people than it used to be.

**Comment by pablo-888** [comment]
Yeah I have noticed that with labwc as well but i'm very much a freak when it comes to latency it's something that always bothered me. I figured most people wouldn't notice and just wouldn't really be bothered by it. I agree that sway seems a little better in that regard but I have no equipment to actually test anything objectively. I just go by feel.

What I do is just enable tearing for all windows (a window rule with * as an identifier) to get the X11 latency experience. I then added another windowrule to specifically disable tearing for mpv. (plus a keybinding to do it on the fly obviously but I don't really use it since watching movies/shows with mpv is the only situation where any tearing would bother me really). 

Not sure if that's a mess of a solution from a developer perspective but i'm more than happy with that setup lol. Plus the only thing that I was missing was the newly merged drawcontents option so i'm pretty much using labwc exclusively now.

**Reply by jp7677** [reply]
Out of curiosity, would https://github.com/labwc/labwc/pull/1568 improve your setup?

**Reply by pablo-888** [reply]
Sorry I'm not sure which part of that relates to my setup unless I explained it poorly (very possible). What I currently have going on looks like this:

	
```
	     <windowRule identifier="*">
        <action name="ToggleTearing"/>
	</windowRule> 
	
		<windowRule identifier="*mpv*" matchOnce="true">
		<action name="ToggleTearing"/>
	</windowRule>
```

- All windows have tearing turned on to reduce input latency (fullscreen, floating, everything).
- Only MPV has it disabled that way I can watch things smoothly without tearing. 

Wouldn't the option in that PR take that away and only enable tearing for things I actually don't want (a fullscreen video player)?

**Reply by jp7677** [reply]
ah, you explained it correctly but I misunderstood :)
I though having that option would get you rid of the window rule to enable tearing everywhere so that you only need to the override for mpv. But since I now understood that you prefer tearing also for windowed applications and not just for (fullscreen) gaming, this indeed doesn't relates to your configuration.

**Reply by tokyo4j** [reply]
> Plus the only thing that I was missing was the newly merged drawcontents option so i'm pretty much using labwc exclusively now.

Oh I haven't recognized there's already a user using `<drawContents>`.

For future development, can I ask you why you need that feature?

While I was making a PR for it, its goal had been a bit unclear to me since the laggy resizing of Chromium and Steam was already fixed with #1861 and I was satisfied without `<drawContents>`.

**Reply by pablo-888** [reply]
I don't use chromium and am not much of a gamer outside of some emulators here and there these days so I wasn't even aware those two had issues resizing 😁.

I'm not sure I *need it* per se but it's just something I've always used when available no matter the OS 

I like the instantaneous/snappy feeling I guess.  

Also for example on older hardware I noticed that most wlroots compositors (especially with vsync on) drop tons of frames when the gtk filechooser has many files or folders in it and you resize the window up to a certain size (/usr/share as an example. not sure if it's a wlroots or a gtk3 wayland thing. I also don't expect most people to run into this to be honest.). 

It's just a peace of mind thing where no matter how big or heavy a program/window/website is you know you're not gonna lag when resizing. Just makes everything feel more pleasant.

**Reply by tokyo4j** [reply]
Great feedback.
I've wondered whether we should add a filled (translucent?) rectangle in addition to outlines as a resizing preview. In your case, I guess a filled rectangle would be helpful to enhance visibility. So I'll consider adding it.
